crd/ensure.go

```diff
@@ -0,0 +1,55 @@
+package crd
+
+import (
+	"context"
+
+	"github.com/cenk/backoff"
+	"github.com/giantswarm/microerror"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func Ensure(ctx context.Context, CRD *CRD, crdClient apiextensionsclient.Interface, backOff backoff.BackOff) error {
+	_, err := crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(CRD.NewResource())
+	if errors.IsAlreadyExists(err) {
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	operation := func() error {
+		manifest, err := crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(CRD.Name(), metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, cond := range manifest.Status.Conditions {
+			switch cond.Type {
+			case apiextensionsv1beta1.Established:
+				if cond.Status == apiextensionsv1beta1.ConditionTrue {
+					return nil
+				}
+			case apiextensionsv1beta1.NamesAccepted:
+				if cond.Status == apiextensionsv1beta1.ConditionFalse {
+					return microerror.Maskf(nameConflictError, cond.Reason)
+				}
+			}
+		}
+
+		return microerror.Mask(notEstablishedError)
+	}
+
+	err = backoff.Retry(operation, backOff)
+	if err != nil {
+		deleteErr := crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(CRD.Name(), nil)
```

Comment:
I don't follow here. If there's an error with ensuring the CRD exists, we delete it?

Reply:
`Ensure` creates the CRD, which is the data object within the k8s API. `Ensure` also verifies if the CRD is populated and excepted. These are k8s internals. There might be conditions that are not accepted and that make the CRD unusable, e.g. in case there are naming conflicts with other resources. So in case the conditions do not meet, we delete it to clean the API. This was all copied from some open source example on how to use CRDs. I did not invent all of this.

Reply:
Okay, that makes sense to me.